### DISCUSSION TITLE
A demonstration program shadows what is to come

### DISCUSSION THREAD

**Discussion by urbanjost** [fortran-lang/fpm#309] [Show and tell]
If you wanted to create a small program that used Regular Expressions, needed some string functions, and should have a ULS interface with long and short options would Fortran seem like the natural language to use.? Well, using `fpm` and sans the commands and help text here is one in ~100 lines that demonstrates how much `fpm` can change what comes naturally when using Fortran. And it is actually useful if you want immediate access to Fortran intrinsic descriptions in your CLI (command line) sessions.

 Introducing 
     [fan](http://github.com/urbanjost/fan)

And if you trust what will happen and have `gfortran`, `git`, and `fpm` installed it can take as little as 
```bash
   git clone https://github.com/urbanjost/fan.git
   cd fan
   fpm build
   fpm install
   fan --help
```
to have it up and running and for altering even if it is your first day using Fortran.

**Comment by everythingfunctional** [comment]
Awesome. This is exactly the kind of thing I was hoping to enable with fpm. You can just pick up a handful of dependencies and throw something together without having to role all your own, or figure out how to build them all together.

**Reply by urbanjost** [reply]
I don't tend to get carried away, but it made me giggle with a little bit of glee myself. Nice to see everything coming together so nicely.

**Comment by ghost** [comment]
Hello. Thank you for this demonstration program. Unfortunately I was not able to run it on Windows. Here is the output.

![build](https://user-images.githubusercontent.com/5913872/102426212-adc37900-3fed-11eb-88e3-b28118d14ecd.png)
☝️ fpm-fortran is unable to download the dependencies. fpm-bootstrap works.

![exec](https://user-images.githubusercontent.com/5913872/102426223-b320c380-3fed-11eb-9dbd-8abd0c947772.png)
☝️ run time error.

**Reply by urbanjost** [reply]
Oops. I just caused the array bounds issue which does not show up with the --release compiler options when I decided I would simplify the code so I just fixed that If you do a "git pull" or  re-clone it that should be repaired. 

 Not pulling the dependency in the Fortran version but in the Haskell/bootstrap version is not something I can reproduce easily; but I suspect others that will see this can. I will look at it, but do not have a similar platform which is probably required to get to the bottom of it.

Thanks once again for working with `fpm`.

**Reply by urbanjost** [reply]
I believe I just got ahead a bit and used syntax in the manifest file (fpm.toml) that the alpha version on the Fortran side was not supporting. Could I ask you try again? I replaced the fpm.toml file.  I was able to duplicate your problem and the new fpm.toml file fixed it on the machine I was on. I hope it fixs it for you too.

**Reply by ghost** [reply]
Hi @urbanjost . I can confirm that fman works on Windows. Thanks!

![fman](https://user-images.githubusercontent.com/5913872/102526801-f70dda00-407a-11eb-8ecf-c26d8b866724.png)

**Comment by urbanjost** [comment]
That was the target!  Hopefully it's actually useful for Fortran users too.  I found it easier to write than in any other language that came to mind, which surprised even me. `fpm` combined with `git` and `github` makes it very easy to distribute and access. And not a parallel numeric vector operation in sight. Hardly feels like Fortran.

**Comment by jvdp1** [comment]
Awesome example, @urbanjost . `fpm` combined with `stdlib` could be awesome and make Fortran really easy to compile and use!

**Reply by urbanjost** [reply]
Exactly what I hope too, although the conditional compilation and a few other issues with stdlib make it a bit difficult to make into an fpm package.  I was thinking it would be easy to take a snapshot of stdlib and make it into an fpm package; but currently that is not as trivial  as I had hoped unless you assume a certain target type.

 I believe that stdlib will be integrated right into fpm, and hopefully into compiler distributions and the Fortran standard itself. In the mean time `fpm` is roaring along and giving everyone a taste of what that will look like.

 Fortran has not been leveraging the Internet nearly enough.  Between good free compilers starting with g95/gfortran and now others and more Fortran packages appearing on repositories like github and projects like stdlib and fpm I hear the rumblings of a renaissance.  Big data and parallel computing are a natural for Fortran and more and more problems require it; with the bar into Fortran lowering and an LLVM version coming it is actually becoming exciting to see all these long-overdue pieces coming together.

In the mean time writing a little app "no one would write in Fortran" and being able to do it more easily than in Python all with tools free on the www that I can get running on a PC is hopefully a proof;  plus I recently started using Mingw and do not have my man-pages so I need this one myself.  I wrote a lot of the man-pages and I still keep needing to look up some of those procedures!

**Reply by jvdp1** [reply]
> Exactly what I hope too, although the conditional compilation and a few other issues with stdlib make it a bit difficult to make into an fpm package. I was thinking it would be easy to take a snapshot of stdlib and make it into an fpm package; but currently that is not as trivial as I had hoped unless you assume a certain target type.

This issue was discussed during the last Fortran Monthly call. As a first step, it was suggested to generate with the CI an archive with all .f90 generated. `fpm` could then load the archive and compile `stdlib`.

**Reply by certik** [reply]
Once this works robustly, `fpm` can even automatically depend on `stdlib`, unless you tell it not to, which will allow people to use Fortran with batteries included.

**Reply by LKedward** [reply]
It's a slightly different approach to how it was discussed at the call, but I've put together an experimental *fpm* package for *stdlib* here:
<https://github.com/LKedward/stdlib-fpm>

Needs a bit of testing, but the included tests run (on Ubuntu) and it seems to work okay as an *fpm* dependency.

**Reply by certik** [reply]
How is it different to what was discussed at the call? It seems you include the generated files. So it seems exactly as it was discussed, the only difference is that you use a dedicated git repository instead of a release tarball. Which improves upon my suggestion and addresses @milancurcic's question how to make it easy to test the latest master.

**Reply by LKedward** [reply]
Yep you're right, the only difference is that it's not a tarball but a separate repo :+1:

**Comment by LKedward** [comment]
Really nice @urbanjost! This will certainly get a lot of use from me - always forgetting the syntax to the lesser used intrinsics! Great demonstration of *fpm*.

**Reply by urbanjost** [reply]
Thanks. I confess to some self-interest in this one as I need to keep looking up those options myself. I think I have written at least a dozen equivalent procedures that were already in the standard that I forgot about so I have a particularly bad need!

**Reply by urbanjost** [reply]
PS:
Although I still use some of my own.  The standard does not require conditioning the data, and I got bit by a SUM() when I used it with ill-conditioned data a long time ago so I still use mine in a private environment.  It bins the data as the data is summed into subranges and has an option to pre-sort the data, which avoided my particular problem, for example. It can make up a lot of the wallclock lost with a parallel sum if there are idle cores around for a decent range of larger problems, etc ... . Hoping to see those kind of options in stdlib.

**Comment by milancurcic** [comment]
I love fman. Thank you for making it. I installed it and I'm sure I'll use it often.

**Reply by urbanjost** [reply]
Great!  I thought it was going to be me and @brocolis. I am starting to rethink agreeing (after persuasive points were made) to taking it out of `fpm`.  Perhaps we can gather small utilities like this into an `fpm-tools` collection. Pre-processors, simple commands to find long lines, expand tabs, convert between dos and unix line terminators, maybe fsplit and f90split (are they PD or MIT licensed?). Some of those are easily available on some platforms (ie. cpp, expand, d2u, dos2unix, ...) and others are not. There are some pretty-printers and reformatters out there but I think (without checking) they have more restrictive licenses (but a list of them might be useful). A little busy right now but __Hmmmm__.

**Reply by certik** [reply]
LFortran can pretty print, but we are still working on improving it: https://gitlab.com/lfortran/lfortran/-/issues?label_name%5B%5D=fmt

And there will be a lot more utilities like that. I would like eventually to have some nice integration with `fpm`, perhaps by installing `fman` or `lfortran`, it can somehow register for `fpm`, so that you can then use it with `fpm`. That is how `lfortran` already registers with `jupyter`.

I created an issue to discuss this more: https://github.com/fortran-lang/fpm/issues/313

**Comment by certik** [comment]
I tried it, the `--help` works:
```
$ fpm run -- --help
 + build/gfortran_debug/app/fman --help
NAME
    fman(1f) - [FUNIX] output descriptions of Fortran intrinsics
    (LICENSE:PD)
...
```
But this fails:
```
$ fpm run -- tan
 + build/gfortran_debug/app/fman "tan"

free(): invalid next size (fast)

Program received signal SIGABRT: Process abort signal.

Backtrace for this error:
#0  0x7f99cafc22ed in ???
#1  0x7f99cafc1503 in ???
#2  0x7f99ca85703f in ???
#3  0x7f99ca856fb7 in ???
#4  0x7f99ca858920 in ???
#5  0x7f99ca8a1966 in ???
#6  0x7f99ca8a89d9 in ???
#7  0x7f99ca8b004f in ???
#8  0x7f99cb12cb3d in ???
#9  0x7f99cb13ec6f in ???
#10  0x7f99cb13edf1 in ???
#11  0x7f99cb396d32 in ???
#12  0x7f99ca85b160 in ???
#13  0x7f99ca85b259 in ???
#14  0x7f99ca839bfd in ???
#15  0x55800ced4379 in ???
#16  0xffffffffffffffff in ???
Aborted (core dumped)
 Command failed
ERROR STOP 

Error termination. Backtrace:
#0  0x45f64d in ???
#1  0x45f8e5 in ???
#2  0x460823 in ???
#3  0x413b98 in __fpm_environment_MOD_run
	at src/fpm_environment.f90:115
#4  0x404b0a in __fpm_MOD_cmd_run
	at src/fpm.f90:424
#5  0x400884 in MAIN__
	at app/main.f90:25
#6  0x400884 in main
	at app/main.f90:8
```
Which reveals several bugs in `fpm`: it should print a nicer stacktrace, it looks like debugging information is missing.

This also fails:
```
$ fpm run 
 + build/gfortran_debug/app/fman 
free(): invalid next size (fast)

Program received signal SIGABRT: Process abort signal.

Backtrace for this error:
#0  0x7f7692d262ed in ???
#1  0x7f7692d25503 in ???
#2  0x7f76925bb03f in ???
#3  0x7f76925bafb7 in ???
#4  0x7f76925bc920 in ???
#5  0x7f7692605966 in ???
#6  0x7f769260c9d9 in ???
#7  0x7f769261404f in ???
#8  0x562597c5c3cd in testit
	at app/fpm_intrinsics.f90:80
#9  0x562597c5d86e in main
	at app/fpm_intrinsics.f90:2
Aborted (core dumped)
 Command failed
ERROR STOP 

Error termination. Backtrace:
#0  0x45f64d in ???
#1  0x45f8e5 in ???
#2  0x460823 in ???
#3  0x413b98 in __fpm_environment_MOD_run
	at src/fpm_environment.f90:115
#4  0x404b0a in __fpm_MOD_cmd_run
	at src/fpm.f90:424
#5  0x400884 in MAIN__
	at app/main.f90:25
#6  0x400884 in main
	at app/main.f90:8
```

I use Ubuntu 18.04.

**Reply by ghost** [reply]
Hi. On Windows, 

```
fpm run -- --help
fpm run -- tan
fpm run
```

work.

**Comment by urbanjost** [comment]
Great! Thanks for testing.

**Reply by urbanjost** [reply]
@certik, what does "gfortran --version" show?  So far I cannot reproduce. There were a lot of things with allocatable CHARACTER variables that did not work as required by the most recent standards till recently.  I am guessing if I use len=80 throughout and make it a little less general it will work from some recent experiences, if the compiler is relatively old.

**Comment by urbanjost** [comment]
I made a quick git(1) push with a kludge patch to test if that is the issue for the fman binary. Is your `fpm` built with debug or release mode?

**Reply by urbanjost** [reply]
Meant to say more clearly if you have a chance could you pull again and see if it resolves the problem with the `fman` executable itself?

**Comment by urbanjost** [comment]
I saw a thread lately talking about the angst and pessimism in Fortran threads. I should point them to some threads like this -- stdlib, fpm, Lfortran, DOE support for gfortran, OpenACC, ...  I see a lot more ambition and optimism than pessimism.

**Comment by urbanjost** [comment]
If anyone likes color, `fman` got renamed to `fan` and has a basic color mode that uses in-line ANSI control sequences, so it does not work in all terminal emulators, but even MSWindows supports it starting with Windows 10; but I think you might have to enable it.
A matter of taste but I like a splash of color.

#  [fan](http://github.com/urbanjost/fan)

**Reply by art-rasa** [reply]
I just installed _fan_ without any problems. I think this is a great utility, especially for beginners like me.

**Comment by urbanjost** [comment]
I am happy you find `fan` useful and that you are using `fpm`. Welcome!  Hopefully, you will also find [Fortran Discourse](https://fortran-lang.discourse.group) and other resources you can find at [fortran-lang](https://github.com/fortran-lang) useful as well for working with Fortran. Thanks for letting me know that it worked for you as well. It's always good to hear feedback like this.
